### PR TITLE
Record-like classes support

### LINKED ISSUE #3082: Does linq2db support C# 9 records?
Hello, does linq2db support C# 9 records?

I am trying to map simple record but get exception `Type 'RecordType' does not have a default constructor`.
Are there any workarounds for records?

Thanks, Ilya.

Example code 

```C#
record Rec(int Id);

[Fact]
public async Task RecordsTest()
{
    var cs = _options.Value.ConnectionString;
    var ms = new MappingSchema();
    ms.GetFluentMappingBuilder().Entity<Rec>()
        .HasTableName("Table")
        .HasSchemaName("dbo")
        .Property(p => p.Id).IsPrimaryKey().HasColumnName("Id").IsIdentity();

    await using var dc = new DataConnection(SqlServerProviderAdapter.MicrosoftProviderFactoryName, cs, ms);
    var data = await dc.GetTable<Rec>().ToArrayAsync();
}
```

Exception stack trace 

```
System.ArgumentException
Type 'Rec' does not have a default constructor (Parameter 'type')
   at System.Linq.Expressions.Expression.New(Type type)
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildDefaultConstructor(EntityDescriptor entityDescriptor, Type objectType, Tuple`2[] index)
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildTableExpression(Boolean buildBlock, Type objectType, Tuple`2[] index)
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildQuery(Type tableType, TableContext tableContext, ParameterExpression parentObject)
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.BuildQuery[T](Query`1 query, ParameterExpression queryParameter)
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]()
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr)
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr)
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache)
   at LinqToDB.Linq.ExpressionQuery`1.GetForEachAsync(Action`1 action, CancellationToken cancellationToken)
   at LinqToDB.AsyncExtensions.ToArrayAsync[TSource](IQueryable`1 source, CancellationToken token)
```

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #3082

### CODE REVIEW

File: Tests/Linq/Linq/MappingTests.cs [hunk 1]
```diff
@@ -533,5 +533,109 @@ public void Issue1592CallbackWithContextConstructor([DataSources] string context
...
+		public abstract class RecordLikeBase
+		{
+			public RecordLikeBase(int Id, string BaseValue)
+			{
+				this.Id = Id;
+				this.BaseValue = BaseValue;
+			}
+
+			public int Id { get; set; }
+			public string BaseValue { get; set; }
```
Comment (Ilchert) [inline Tests/Linq/Linq/MappingTests.cs:562]: I think, you are missing [`init only properties`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/init) test case.

Comment (sdanyliv) [reply Tests/Linq/Linq/MappingTests.cs]: Just tested, it works in this case. Corrected anyway.

File: Source/LinqToDB/Linq/Builder/TableBuilder.TableContext.cs [hunk 2]
```diff
@@ -264,6 +264,19 @@ bool IsAnonymous(Type type)
 				}
 
 				return false;
+			}			
+			
+			bool HasDefaultConstructor(Type type)
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Builder/TableBuilder.TableContext.cs:269]: do we need to check for constructor access too?

**Pull request merged**